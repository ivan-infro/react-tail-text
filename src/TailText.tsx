import React, { MouseEventHandler, useEffect, useRef, useMemo } from "react";
import styled from "styled-components";
import memoize from "lodash.memoize";
import addToObserver from "./observer";
import {
  fulltextClassName,
  tailClassName,
  wrapperClassName,
  wrapperTailedClassName,
} from "./constants";
import { InvisibleTail } from "./InvisibleTail";

type TailTextPros = {
  /**
   * Текст, который необходимо вписать по ширине
   */
  children: string | number;

  /**
   * Количество символов в конце строки, видимых всегда
   *
   * Должно быть не больше длины строки props.children
   */
  tailLength: number;

  /**
   * Подсказка, отображающася при наведении
   */
  title?: string;

  /**
   * Дополнительный className для элемента-обертки
   */
  className?: string;
};

const selectFulltext: MouseEventHandler<HTMLDivElement> = (event) => {
  if (typeof window === "undefined") return;
  window.getSelection()?.selectAllChildren(event.currentTarget);
};

const getTail = memoize((fulltext: number | string, tailLength: number) => {
  const str = fulltext.toString();
  if (tailLength > str.length) {
    console.warn("tailLength не может быть больше длины исходной строки!");
  }
  return str.substring(str.length - tailLength);
});

const StyledTailText = styled(TailText)`
  white-space: nowrap;
  max-width: 100%;
  overflow: hidden;

  &:hover {
    color: red;
  }

  &:focus-within {
    color: blue;
  }

  &:active {
    color: purple;
  }

  &.${wrapperTailedClassName}::after {
    content: "${(props) => getTail(props.children, props.tailLength)}";
  }
`;

type FulltextProps = {
  tailWidth?: number;
  className?: string;
  children: string | number;
};

const Fulltext = (props: FulltextProps) => {
  return <div {...props}>{props.children}</div>;
};

const StyledFulltext = styled(Fulltext)`
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  vertical-align: top;

  &.react-tail-text__fulltext_selected {
    background: yellow;
  }
`;

function TailText(props: TailTextPros) {
  const wrapperRef = useRef<HTMLDivElement>(null);

  const tail = getTail(props.children, props.tailLength);

  useEffect(() => {
    if (!wrapperRef.current) {
      console.warn("wrapperRef.current is null!");
      return;
    }
    return addToObserver(wrapperRef.current);
  }, []);

  const className = useMemo(
    () => [wrapperClassName, props.className].join(" "),
    [props.className]
  );

  return (
    <div
      className={className}
      title={props.title}
      ref={wrapperRef}
      onDoubleClick={selectFulltext}
    >
      <InvisibleTail className={tailClassName}>{tail}</InvisibleTail>
      <StyledFulltext className={fulltextClassName}>
        {props.children}
      </StyledFulltext>
    </div>
  );
}

export default StyledTailText;
