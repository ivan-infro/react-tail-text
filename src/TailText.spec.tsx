import React from "react";
import renderer from "react-test-renderer";
import TailText from "./TailText";

test("it is alive", () => {
  const tree = renderer
    .create(<TailText tailLength={3}>it is alive</TailText>)
    .toJSON();

  expect(tree).toMatchSnapshot();

  const secondTree = renderer
  .create(<TailText tailLength={1}>да будет свет</TailText>)
  .toJSON();

expect(secondTree).toMatchSnapshot();
});

test("Текст подскази при наведении добавлен в атрибут title", () => {
  const tree = renderer
    .create(
      <TailText tailLength={3} title="А вот и сама подсказка">
        Подсказка при наведении добавлена к wrapper элементу в тег title
      </TailText>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("Выводится предупреждение о некорректном параметре teilLength", () => {
  const spy = jest.spyOn(console, "warn").mockImplementation();
  const tree = renderer
    .create(<TailText tailLength={100500}>Краткость - сестра таланта</TailText>)
    .toJSON();

  expect(tree).toMatchSnapshot();
  expect(spy).toHaveBeenCalledTimes(2);
  expect(spy).toHaveBeenLastCalledWith(
    "tailLength не может быть больше длины исходной строки!"
  );
  spy.mockRestore();
});

test("Кастомный className добавляется к элементу", () => {
  const tree = renderer
    .create(
      <TailText tailLength={3} className="hello">
        Модифицировать вшитые классы пока нельзя
      </TailText>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("children - это не только ценный актив, но и три - четыре... Ну, и другие числа", () => {
  const tree = renderer.create(<TailText tailLength={1}>34</TailText>).toJSON();

  expect(tree).toMatchSnapshot();
});
