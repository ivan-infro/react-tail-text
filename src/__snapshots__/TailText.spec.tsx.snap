// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`children - это не только ценный актив, но и три - четыре... Ну, и другие числа 1`] = `
.c1 {
  visibility: hidden;
  width: 0;
  height: 0;
  white-space: nowrap;
  overflow: hidden;
}

.c0 {
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
}

.c0.react-tail-text_tailed::after {
  content: "4";
}

.c2 {
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  vertical-align: top;
}

.c2.react-tail-text__fulltext_selected {
  background: yellow;
}

<div
  className="react-tail-text c0"
  onDoubleClick={[Function]}
>
  <div
    className="c1 react-tail-text__tail"
  >
    4
  </div>
  <div
    className="c2 react-tail-text__fulltext"
  >
    34
  </div>
</div>
`;

exports[`it is alive 1`] = `
.c1 {
  visibility: hidden;
  width: 0;
  height: 0;
  white-space: nowrap;
  overflow: hidden;
}

.c0 {
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
}

.c0.react-tail-text_tailed::after {
  content: "ive";
}

.c2 {
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  vertical-align: top;
}

.c2.react-tail-text__fulltext_selected {
  background: yellow;
}

<div
  className="react-tail-text c0"
  onDoubleClick={[Function]}
>
  <div
    className="c1 react-tail-text__tail"
  >
    ive
  </div>
  <div
    className="c2 react-tail-text__fulltext"
  >
    it is alive
  </div>
</div>
`;

exports[`it is alive 2`] = `
.c1 {
  visibility: hidden;
  width: 0;
  height: 0;
  white-space: nowrap;
  overflow: hidden;
}

.c0 {
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
}

.c0.react-tail-text_tailed::after {
  content: "т";
}

.c2 {
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  vertical-align: top;
}

.c2.react-tail-text__fulltext_selected {
  background: yellow;
}

<div
  className="react-tail-text c0"
  onDoubleClick={[Function]}
>
  <div
    className="c1 react-tail-text__tail"
  >
    т
  </div>
  <div
    className="c2 react-tail-text__fulltext"
  >
    да будет свет
  </div>
</div>
`;

exports[`Выводится предупреждение о некорректном параметре teilLength 1`] = `
.c1 {
  visibility: hidden;
  width: 0;
  height: 0;
  white-space: nowrap;
  overflow: hidden;
}

.c0 {
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
}

.c0.react-tail-text_tailed::after {
  content: "Краткость - сестра таланта";
}

.c2 {
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  vertical-align: top;
}

.c2.react-tail-text__fulltext_selected {
  background: yellow;
}

<div
  className="react-tail-text c0"
  onDoubleClick={[Function]}
>
  <div
    className="c1 react-tail-text__tail"
  >
    Краткость - сестра таланта
  </div>
  <div
    className="c2 react-tail-text__fulltext"
  >
    Краткость - сестра таланта
  </div>
</div>
`;

exports[`Кастомный className добавляется к элементу 1`] = `
.c1 {
  visibility: hidden;
  width: 0;
  height: 0;
  white-space: nowrap;
  overflow: hidden;
}

.c0 {
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
}

.c0.react-tail-text_tailed::after {
  content: "ьзя";
}

.c2 {
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  vertical-align: top;
}

.c2.react-tail-text__fulltext_selected {
  background: yellow;
}

<div
  className="react-tail-text c0 hello"
  onDoubleClick={[Function]}
>
  <div
    className="c1 react-tail-text__tail"
  >
    ьзя
  </div>
  <div
    className="c2 react-tail-text__fulltext"
  >
    Модифицировать вшитые классы пока нельзя
  </div>
</div>
`;

exports[`Текст подскази при наведении добавлен в атрибут title 1`] = `
.c1 {
  visibility: hidden;
  width: 0;
  height: 0;
  white-space: nowrap;
  overflow: hidden;
}

.c0 {
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
}

.c0.react-tail-text_tailed::after {
  content: "tle";
}

.c2 {
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  vertical-align: top;
}

.c2.react-tail-text__fulltext_selected {
  background: yellow;
}

<div
  className="react-tail-text c0"
  onDoubleClick={[Function]}
  title="А вот и сама подсказка"
>
  <div
    className="c1 react-tail-text__tail"
  >
    tle
  </div>
  <div
    className="c2 react-tail-text__fulltext"
  >
    Подсказка при наведении добавлена к wrapper элементу в тег title
  </div>
</div>
`;
